src/hyperlight_common/src/arch/amd64/vm.rs

```diff
@@ -0,0 +1,217 @@
+) -> ModifyPteIterator<HIGH_BIT, LOW_BIT, Op> {
+    ModifyPteIterator { request: r, n: 0 }
+}
+
+/// Page-mapping callback to allocate a next-level page table if necessary.
+/// # Safety
+/// This function modifies page table data structures, and should not be called concurrently
+/// with any other operations that modify the page tables.
+unsafe fn alloc_pte_if_needed<Op: TableOps>(
+    op: &Op,
+    x: MapResponse<Op::TableAddr>,
+) -> MapRequest<Op::TableAddr> {
+    let pte = unsafe { op.read_entry(x.entry_ptr) };
+    let present = pte & 0x1;
+    if present != 0 {
+        return MapRequest {
+            table_base: Op::from_phys(pte & !0xfff),
+            vmin: x.vmin,
+            len: x.len,
+        };
+    }
+
+    let page_addr = unsafe { op.alloc_table() };
+
+    #[allow(clippy::identity_op)]
+    #[allow(clippy::precedence)]
+    let pte = Op::to_phys(page_addr) |
+        1 << 5 | // A   - we don't track accesses at table level
+        0 << 4 | // PCD - leave caching enabled
+        0 << 3 | // PWT - write-back
+        1 << 2 | // U/S - allow user access to everything (for now)
+        1 << 1 | // R/W - we don't use block-level permissions
+        1 << 0; // P   - this entry is present
+    unsafe { op.write_entry(x.entry_ptr, pte) };
+    MapRequest {
+        table_base: page_addr,
+        vmin: x.vmin,
+        len: x.len,
+    }
+}
+
+/// Map a normal memory page
+/// # Safety
+/// This function modifies page table data structures, and should not be called concurrently
+/// with any other operations that modify the page tables.
+#[allow(clippy::identity_op)]
+#[allow(clippy::precedence)]
+unsafe fn map_page<Op: TableOps>(op: &Op, mapping: &Mapping, r: MapResponse<Op::TableAddr>) {
+    let pte = match &mapping.kind {
+        MappingKind::BasicMapping(bm) =>
+        // TODO: Support not readable
+        {
+            (mapping.phys_base + (r.vmin - mapping.virt_base)) |
+                (!bm.executable as u64) << 63 | // NX - no execute unless allowed
+                1 << 7 | // 1   - RES1 according to manual
+                1 << 6 | // D   - we don't presently track dirty state for anything
+                1 << 5 | // A   - we don't presently track access for anything
+                0 << 4 | // PCD - leave caching enabled
+                0 << 3 | // PWT - write-back
+                1 << 2 | // U/S - allow user access to everything (for now)
+                (bm.writable as u64) << 1 | // R/W - for now make everything r/w
+                1 << 0 // P   - this entry is present
+        }
+    };
+    unsafe {
+        op.write_entry(r.entry_ptr, pte);
+    }
+}
+
+// There are no notable architecture-specific safety considerations
+// here, and the general conditions are documented in the
+// architecture-independent re-export in vm.rs
+#[allow(clippy::missing_safety_doc)]
+pub unsafe fn map<Op: TableOps>(op: &Op, mapping: Mapping) {
+    modify_ptes::<47, 39, Op>(MapRequest {
```

Comment:
I know this is code that was just moved around, but it would be nice if there was some comments explaining what's going on here.